nexus/reconfigurator/planning/src/blueprint_builder.rs

```diff
@@ -1025,6 +1165,26 @@ mod builder_zones {
             Ok(())
         }
 
+        pub(super) fn expunge_zones(
+            &mut self,
+            zones: &[OmicronZoneUuid],
+        ) -> Result<(), BuilderZonesConfigError> {
+            for zone_id in zones {
+                // A linear search for each zone is fine, each sled has a pretty
+                // small number of zones.
+                if let Some(zone) =
+                    self.zones.iter_mut().find(|z| z.zone.id == *zone_id)
+                {
+                    // Just check that the zone is still expungeable.
+                    is_already_expunged(&zone.zone, zone.state)?;
```

Comment:
I don't entirely follow what this is guarding here. In particular, what would a return value of `Ok(true)` mean? We already expunged the zone but forgot to mark the state as `Modified` somehow?

Reply:
This is making sure that any zones passed in at this level are in a state where marking them as expunged is valid. This is an unavoidable check that happens a second time. The list of zones to expunge has to be passed in -- it cannot be constructed here, because a `BuilderZonesConfig` is only created after we've decided to change the sled. If there are no zones to expunge then a `BuilderZonesConfig` shouldn't be created.

Reply:
Hm, but am I right that `Ok(true)` would be a nonsensical return value? Would it make sense to separate "validate that this zone is expungeable" from "has the zone already been marked expunged", and only call the first one here?

Reply:
`Ok(true)` indicates a no-op (idempotence). I changed the definition of `is_already_expunged` slightly, and especially with this new definition -- but even with the old one -- I don't think it makes sense to split them out. The simplest definition of a `validate_expungeable` function would be to just call `is_already_expunged()?;` and drop the return value.

(I went through this series of arguments myself while writing this code. Options I considered included renaming `is_already_expunged` to also include validation -- I decided that was too wordy, and already indicated by the fact that it returns a `Result`.)